compiler/il/Aliases.cpp

```diff
@@ -105,6 +105,20 @@ OMR::SymbolReference::getUseonlyAliasesBV(TR::SymbolReferenceTable * symRefTab)
       case TR::Symbol::IsMethod:
          {
          TR::MethodSymbol * methodSymbol = _symbol->castToMethodSymbol();
+
+         // No aliasing is needed for potentialOSRPointHelper and osrFearPointHelper because:
+         // The two helper calls are considered exception points because the Node API exceptionsRaised
+         // returns non-zero value for any call. The control flow constraint imposed by the exception
+         // edge will apply to all the global optimizations that may move things around. Local
+         // optimizations also ask `exceptionsRaised` to determine if a movement across certain point
+         // is dangerous.
```

Comment:
So I buy that this would work for the fear points the way they are intended to be used - that is we create a fear point targeting a bytecode index in a given method when we believe the block already has the infrastructure to do that transition. I am not sure that this will hold for a fear point. The fear point is not a point that we explicitly kept track of during ilgen and so you are relying on the block's OSR path to keep things live for a point that was not tracked during ilgen. Now if the fear points are also potential OSR points then I can see how that would work, but there is an implicit assumption on the relationship between the fear point and the potential points.

How can we be sure the exception edge for the fear point a) will exist and b) will hold all the right things live?

Reply:
The exception control flow is relevant for the OSR points, but not (directly) for the fear points.

Because a fear point helper call must be protected, and because it represents no computation of its own, I believe the only constraints on it have to do with this protection. If it's unprotectable, it must _remain_ unprotectable. If OTOH it's protectable, the set of OSR points protecting it must be (roughly speaking) preserved. I say roughly speaking because any such OSR point that isn't a yield point (e.g. a potential OSR point helper call), if it could itself be protected by earlier OSR points, can be replaced with those. Protectable fear point helper calls are also allowed to become unprotectable, although that's undesirable because it would cause spurious compilation restarts.

I think these invariants could be maintained while allowing reordering of fear point helper calls w.r.t. anything other than OSR points and yield/invalidation points. These reorderings would in almost all cases be naturally disallowed simply due to the fact that the fear point is represented as a call, which even without aliasing could e.g. perform I/O.

That said, I think it's perfectly fine to add aliasing here to cause the compiler to be more conservative about reordering these helpers with other operations, so long as there is a comment to indicate that the aliasing is known to be more conservative than necessary. That would hedge against any holes in the account I just gave, and it can always be relaxed later if necessary.